cpp/include/cudf/io/text/superstate.hpp

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include <cmath>
+#include <cstdint>
+#include <type_traits>
+
+namespace {
+
+constexpr unsigned floorlog2(unsigned x) { return x == 1 ? 0 : 1 + floorlog2(x >> 1); }
+
+constexpr unsigned ceillog2(unsigned x) { return x == 1 ? 0 : floorlog2(x - 1) + 1; }
+
+template <uint8_t Bits, typename Enable = void>
+struct rep {
+};
+
+template <uint8_t Bits>
+struct rep<Bits, std::enable_if_t<0 < Bits and Bits <= 8>> {
+  using type = uint8_t;
+};
+
+template <uint8_t Bits>
+struct rep<Bits, std::enable_if_t<8 < Bits and Bits <= 16>> {
```

Comment:
have not experimented with it. but at the time, I was thinking space was important, too, since it could be used as tile state. I do recall getting weird ptx when testing in godbolt with smaller types, though.